src/som/compiler/MethodBuilder.java

```diff
@@ -215,17 +209,17 @@ public boolean requiresContext() {
   }
 
   private MethodBuilder markOuterContextsToRequireContextAndGetRootContext() {
-    MethodBuilder ctx = outerBuilder;
-    while (ctx.outerBuilder != null) {
+    MethodBuilder ctx = getNextMethod();
+    while (ctx.getNextMethod() != null) {
       ctx.throwsNonLocalReturn = true;
-      ctx = ctx.outerBuilder;
+      ctx = ctx.getNextMethod();
     }
     return ctx;
   }
 
   public boolean needsToCatchNonLocalReturn() {
     // only the most outer method needs to catch
-    return needsToCatchNonLocalReturn && outerBuilder == null;
+    return needsToCatchNonLocalReturn && getContextualSelf() instanceof MixinBuilder;
```

Comment:
Do you mind if I change my mind on stuff?

We might indeed consider `isMixin()` and `isMethod()` methods to avoid all these `instanceof` tests.